### PR TITLE
[TTS] Add Mandarin/English Bilingual Recipe for Training Fastpitch Models

### PR DESCRIPTION
**Pull request by yuekaizhang**
Signed-off-by: Yuekai Zhang <zhangyuekai@foxmail.com>

Add Chinese TTS support with bilingual dataset [SF_bilingual_dataset](https://catalog.ngc.nvidia.com/orgs/nvidia/resources/sf_bilingual_speech_zh_en)

Closed previous #5197 to modify the sign-off info.

@XuesongYang Sorry for the duplicate PR, would you mind reviewing it again? Thank you.

### CODE REVIEW

**Comment by XuesongYang** [comment]
no worries, please. I just realized the previous one was a draft. Let me know once this PR is ready to review. Thanks!

**Comment by yuekaizhang** [comment]
> no worries, please. I just realized the previous one was a draft. Let me know once this PR is ready to review. Thanks!

Hi Xuesong, I think this PR could be reviewed now. Btw, do I need to put the pretrained model somewhere?

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py
Comment (XuesongYang) [inline nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:736]: could you pls make an OOV example of Chinese?

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:511]: typo: `pinyin2phonemes` should be consistent with `pinyin_dict_nv22.10`.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:514]: typo: pinyin2phonemes should be consistent with pinyin_dict_nv22.10.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:525]: `heteronyms`, `encoding` and `phoneme_probability` params are not used as inputs to `__init__`. Is there any reason why we decide to remove them in `self.__init__()`?

IIUC, after conversion into phonemes from Chinese characters, the heteronyms should not exist, so that `phoneme_probability` is always 1.0. So maybe that's why you remove them?

Comment (XuesongYang) [reply nemo_text_processing/g2p/modules.py:525]: A general question: how to address the case that OOV characters do not exist in pypinyin?

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:570]: it seems that individual English letters are used as "phonemes", while g2p did not look English words up in the English dict to get English phonemes. So the synthesized audio of English words is just like letter sound, right? 

If so, is there any way to improve it by looking up our English dict?

File: scripts/dataset_processing/tts/sfbilingual/ds_conf/ds_for_fastpitch_align.yaml
Comment (XuesongYang) [inline scripts/dataset_processing/tts/sfbilingual/ds_conf/ds_for_fastpitch_align.yaml:25]: I would suggest applying the same trim config as `fastpitch_align_22050.yaml` above, otherwise, it may diverge the pitch norm slightly.

**Review by XuesongYang — CHANGES_REQUESTED** [review]
Thanks for your instant PR! I left some comments above. Please have a look.

File: scripts/tts_dataset_files/zh/pinyin_dict.txt
Comment (yuekaizhang) [reply scripts/tts_dataset_files/zh/pinyin_dict.txt:1]: Sure, I have sorted it.

File: examples/tts/conf/zh/fastpitch_align_22050.yaml
Comment (yuekaizhang) [reply examples/tts/conf/zh/fastpitch_align_22050.yaml:92]: No worry, I change it to null.

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py
Comment (yuekaizhang) [reply nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:736]: Sorry, I just copied comments from EnglishPhoneTokenier. There is almost no OOV for Chinese. Pypinyin covers 99.9999% Chinese charaters.

File: nemo_text_processing/g2p/modules.py
Comment (yuekaizhang) [reply nemo_text_processing/g2p/modules.py:525]: Yes, I have no consider for polyphone disambiguation in this PR. Maybe it could be done through a unified frontend along with a prosody bert. 

For OOV cases in Chinese, I think pypinyin cover almost all of it. (Chinese characters are standord fixed limited set) In case it can't handle it, it would return the token unchanged. In the following ChinesePhoneTokenier, the out of dict token would be ignored with a warning information.

File: nemo_text_processing/g2p/modules.py
Comment (yuekaizhang) [reply nemo_text_processing/g2p/modules.py:570]: Not exactly. For Chinese, the model uses phone as modeling unit. While for English, the model will receive grapheme representation of text (letters) as input. What's the synthesized audio sounds like would depend on the ground truth audio. E.g. for "cat", it could be "c ae t" or "c a t".

It is possible to use phone for English modeling unit. However, I chose to use letters for code simplification, since we need to combine pureChinesePhoneTokenier with pureEnglishPhoneTokenier into a bilingual Tokenier. (EnglishPhoneTokenier is much more complicated than EnglishCharTokenier)

**Comment by yuekaizhang** [comment]
@XuesongYang, many thanks for your review. I have modified the codes according the comments.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [reply nemo_text_processing/g2p/modules.py:565]: > What's the synthesized audio sounds like would depend on the ground truth audio. E.g. for "cat", it could be "c ae t" or "c a t".

understood. We are good for now. The only concern is that parallel English letters and audio clips are rather scarce in the bilingual setup compared to the rich Chinese data resources. Look forward to the listening test and verifying it further.

File: examples/tts/conf/zh/fastpitch_align_22050.yaml
Comment (XuesongYang) [reply examples/tts/conf/zh/fastpitch_align_22050.yaml]: thanks. But please make sure OOM doesn't happen during training.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [reply nemo_text_processing/g2p/modules.py]: sounds good!

File: examples/tts/conf/zh/fastpitch_align_22050.yaml
Comment (yuekaizhang) [reply examples/tts/conf/zh/fastpitch_align_22050.yaml]: Of course, I have checked all audios are shorter than 10 secs. Since this dataset only contains 4.5 hours, there is not necessary to discard long audios.

File: nemo_text_processing/g2p/modules.py
Comment (yuekaizhang) [reply nemo_text_processing/g2p/modules.py:565]: Yes. I think for this bilingual dataset, it would be better to start from a LJspeech pretrained model which covers more English. Maybe consider to add a notebook tutorial for this two stage training (LJspeech -> SF_bilingual) in next PR.

I would prepare some audios and ground truth audios for check.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:548]: pls remove `self` and add a decor `@staticmethod` to this func cuz `self` is not used anywhere in this func.

File: scripts/dataset_processing/tts/sfbilingual/ds_conf/ds_for_fastpitch_align.yaml
Comment (XuesongYang) [inline scripts/dataset_processing/tts/sfbilingual/ds_conf/ds_for_fastpitch_align.yaml:7]: pls correct the new name of dict.

File: scripts/dataset_processing/tts/sfbilingual/get_data.py
Comment (XuesongYang) [inline scripts/dataset_processing/tts/sfbilingual/get_data.py:70]: Does this `TODO` need to be addressed soon? Not quite sure which pattern is used for the final replace.

**Review by XuesongYang — CHANGES_REQUESTED** [review]

**Comment by XuesongYang** [comment]
Thanks for the quick updates. I added some neat-picks. I am running through your tutorial, and will get back to you tomorrow.

File: nemo_text_processing/g2p/modules.py
Comment (yuekaizhang) [reply nemo_text_processing/g2p/modules.py:543]: I thought people who don't need to use Chinese TTS could ignore pypinyin. I am good to put it at the begining of the file.

**Comment by XuesongYang** [comment]
@redoctopus @ekmb could you please have a look at this PR, especially G2P and tokenizer part? It looks good to me.

File: scripts/dataset_processing/tts/sfbilingual/get_data.py
Comment (yuekaizhang) [reply scripts/dataset_processing/tts/sfbilingual/get_data.py:70]: It's a mistake of dataset itself. Wav's name is like com_SF_0001.wav, however, in the label file they offered, wav's name is com_DL_00001. When they fix it in the dataset, we could remove this WAR here.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [reply nemo_text_processing/g2p/modules.py:543]: good point. Let's keep as-is.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:507]: `ignore_ambiguous_words` is not called in this func. Pls remove it.

File: nemo_text_processing/g2p/modules.py
Comment (XuesongYang) [inline nemo_text_processing/g2p/modules.py:520]: remove docstring for `ignore_ambiguous_words`.

File: scripts/dataset_processing/tts/sfbilingual/get_data.py
Comment (XuesongYang) [reply scripts/dataset_processing/tts/sfbilingual/get_data.py:70]: are they working on the fix? if not, could you please add such detailed comments there? Thanks.

Comment (yuekaizhang) [reply scripts/dataset_processing/tts/sfbilingual/get_data.py:70]: I would let them know. For now, add more details in the comments.

File: tutorials/tts/FastPitch_ChineseTTS_Training.ipynb
Comment (XuesongYang) [inline tutorials/tts/FastPitch_ChineseTTS_Training.ipynb:922]: need to add a `\` at the end of Line 922, otherwise, indent error happens.

``` bash
!(cd NeMoChineseTTS && \
    python get_data.py \
        --data-root ./DataChinese/sf_bilingual_speech_zh_en_vv1/SF_bilingual/ \
        --manifests-path ./)
```

File: tutorials/tts/FastPitch_ChineseTTS_Training.ipynb
Comment (XuesongYang) [inline tutorials/tts/FastPitch_ChineseTTS_Training.ipynb:178]: need a line `cd sf_bilingual_speech_zh_en_vv1 && \`, otherwise, error occurs.

**Review by XuesongYang — CHANGES_REQUESTED** [review]
the tutorial seems broken to me. Could you please make sure it can be running through without errors? Thanks.

**Comment by yuekaizhang** [comment]
> the tutorial seems broken to me. Could you please make sure it can be running through without errors? Thanks.

Sorry, my bad. Would run it from scratch and let you know when I finish it. Maybe you could review it later.

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py
Comment (yuekaizhang) [reply nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:648]: I think perhaps we could remove it. TN zh would convert English punct into Chinese punct.

File: nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py
Comment (yuekaizhang) [reply nemo/collections/common/tokenizers/text_to_speech/tts_tokenizers.py:679]: For English, we choose letter inputs. I think I could remove it.

File: nemo_text_processing/g2p/modules.py
Comment (yuekaizhang) [reply nemo_text_processing/g2p/modules.py:519]: Sorry, I think phoneme_dict here can't be None. Would remove it.

**Comment by yuekaizhang** [comment]
@XuesongYang , would you mind checking the notebook again?

Also attached two audios and ground truth (all used pretrained hifi-gan from LJspeech)
https://drive.google.com/drive/folders/1LpMKg-yCfshpOBiyRYPRGPBusnWuUOsS?usp=sharing

**Review by redoctopus — COMMENTED** [review]
Tokenizer & G2P looks good to me! Will wait on @XuesongYang's approval.

**Review by rlangman — COMMENTED** [review]
Should we be pushing a new tutorial which is almost identical/copied from the German tutorial?

I think the tutorial and get_data script/config should be reviewed as a separate PR from the text processing code.

Also, please clear the cell outputs of the tutorials/notebooks before uploading. The outputs make the notebook too large to view and comment on in github.

**Comment by XuesongYang** [comment]
> Should we be pushing a new tutorial which is almost identical/copied from the German tutorial?

Thank you!

**Comment by XuesongYang** [comment]
> Tokenizer & G2P looks good to me! Will wait on @XuesongYang's approval.

Thank you for the safeguard!

**Comment by XuesongYang** [comment]
> Should we be pushing a new tutorial which is almost identical/copied from the German tutorial?

That's the plan that we would unify the tutorial for any languages and suitable model architectures. We can merge this PR for now and refactor later.

> I think the tutorial and get_data script/config should be reviewed as a separate PR from the text processing code.

Literally, yes. But for the workload of this PR, it is fine. We can test the whole pipeline in a PR.

> Also, please clear the cell outputs of the tutorials/notebooks before uploading. The outputs make the notebook too large to view and comment on in github.

agreed. @yuekaizhang could you pls clear outputs of tutorial and update again? Thanks.

**Comment by yuekaizhang** [comment]
> update again? Thanks.

Sure. Let me clean it.

**Review by XuesongYang — CHANGES_REQUESTED** [review]
LGTM. Let's make minor change before merging

File: examples/tts/conf/zh/fastpitch_align_22050.yaml
Comment (yuekaizhang) [reply examples/tts/conf/zh/fastpitch_align_22050.yaml:19]: Yes, they are computed from SF.

**Review by XuesongYang — APPROVED** [review]

**Review by XuesongYang — APPROVED** [review]
LGTM. Thanks for your contribution!

**Pull request merged**
